### PR TITLE
adds claw machines

### PR DESCRIPTION
**Pull request by boodaliboo**
ports https://github.com/ParadiseSS13/Paradise/pull/2796
and https://github.com/ParadiseSS13/Paradise/pull/2788
i spent 3 hours on this
adds a new html5 javascript claw game its cool
https://camo.githubusercontent.com/9651d3219167dd495e883fd493081dff28459b06/68747470733a2f2f692e6779617a6f2e636f6d2f34336133356536373338373661323561353435396539383830333663376339312e676966

#### Changelog

:cl:  
rscadd: Added new arcade game 
/:cl:

### CODE REVIEW

**Comment by boodaliboo** [comment]
i need help please

**Comment by boodaliboo** [comment]
help

File: code/modules/arcade/arcade_base.dm [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+		if(src != user.machine)
+			user << "Someone else is already playing this machine, please wait your turn!"
+		return
+	if(playing && (src != user.machine))
+		user << "Someone else is already playing this machine, please wait your turn!"
+		return 0
+	return 1
+
+/obj/machinery/arcade/attackby(var/obj/item/O as obj, var/mob/user as mob, params)
+	if(istype(O, /obj/item/weapon/screwdriver) && anchored)
```
Comment (missatessatessy) [inline code/modules/arcade/arcade_base.dm]: /obj/item/screwdriver

File: code/modules/arcade/arcade_base.dm [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+		playsound(src.loc, 'sound/items/Screwdriver.ogg', 50, 1)
+		panel_open = !panel_open
+		user << "You [panel_open ? "open" : "close"] the maintenance panel."
+		update_icon()
+		return
+	if(!freeplay)
+		if(istype(O, /obj/item/weapon/card/id))
+			var/obj/item/weapon/card/id/C = O
```
Comment (missatessatessy) [inline code/modules/arcade/arcade_base.dm]: /obj/item/card/id
for 73 too

File: code/modules/arcade/arcade_base.dm [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+			if(pay_with_card(C))
+				tokens += 1
+			return
+		else if(istype(O, /obj/item/weapon/spacecash))
+			var/obj/item/weapon/spacecash/C = O
```
Comment (missatessatessy) [inline code/modules/arcade/arcade_base.dm]: /obj/item/stack/spacecash
seriously, stop calling things weapons

File: code/modules/arcade/arcade_base.dm [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+			if(pay_with_cash(C, user))
+				tokens += 1
+			return
+	if(panel_open&& component_parts && istype(O, /obj/item/weapon/crowbar))
+		default_deconstruction_crowbar(O)
+
+/obj/machinery/arcade/update_icon()
+	return
+
+/obj/machinery/arcade/proc/pay_with_cash(var/obj/item/weapon/spacecash/cashmoney, var/mob/user)
```
Comment (missatessatessy) [inline code/modules/arcade/arcade_base.dm]: /obj/item/stack/spacecash

File: code/modules/arcade/arcade_base.dm [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+			var/obj/item/weapon/card/id/C = O
+			if(pay_with_card(C))
+				tokens += 1
+			return
+		else if(istype(O, /obj/item/weapon/spacecash))
+			var/obj/item/weapon/spacecash/C = O
+			if(pay_with_cash(C, user))
+				tokens += 1
+			return
+	if(panel_open&& component_parts && istype(O, /obj/item/weapon/crowbar))
```
Comment (missatessatessy) [inline code/modules/arcade/arcade_base.dm]: /obj/item/crowbar

File: code/modules/arcade/arcade_base.dm [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+		return 0
+	visible_message("<span class='info'>[usr] inserts a credit chip into [src].</span>")
+	var/left = cashmoney.get_total() - token_price
+	user.unEquip(cashmoney)
+	qdel(cashmoney)
+	if(left)
+		dispense_cash(left, src.loc, user)
+	return 1
+
+/obj/machinery/arcade/proc/pay_with_card(var/obj/item/weapon/card/id/I, var/mob/user)
```
Comment (missatessatessy) [inline code/modules/arcade/arcade_base.dm]: not a weapon again

File: code/modules/arcade/claw_game.dm [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+/obj/machinery/arcade/claw/New()
+	src.addAtProcessing()
+
+	machine_image = pick("_1", "_2")
+	update_icon()
+	
+	component_parts = list()
+	component_parts += new /obj/item/weapon/circuitboard/clawgame(null)
+	component_parts += new /obj/item/weapon/stock_parts/matter_bin(null)
+	component_parts += new /obj/item/weapon/stock_parts/manipulator(null)
```
Comment (missatessatessy) [inline code/modules/arcade/claw_game.dm]: none of these have weapon in their type paths

File: code/modules/arcade/claw_game.dm [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+	component_parts += new /obj/item/stack/sheet/glass(null, 1)
+	RefreshParts()
+	
+	if(!claw_game_html)
+		claw_game_html = file2text('code/modules/arcade/crane.html')
+		claw_game_html = replacetext(claw_game_html, "/* ref src */", "\ref[src]")
+
+/obj/machinery/arcade/claw/RefreshParts()
+	var/bin_upgrades = 0
+	for(var/obj/item/weapon/stock_parts/matter_bin/B in component_parts)
```
Comment (missatessatessy) [inline code/modules/arcade/claw_game.dm]: weapon

File: code/modules/research/designs/comp_board_designs.dm [hunk 9]
```diff
@@ -22,6 +22,14 @@
 	category = list("Computer Boards")
 	departmental_flags = DEPARTMENTAL_FLAG_ALL
 
+/datum/design/clawgame
+	name = "Machine Design (Claw Game Board)"
+	desc = "The circuit board for a Claw Game."
+	id = "clawgame"
+	build_type = IMPRINTER
+	category = list("Computer Boards")
+	build_path = /obj/item/weapon/circuitboard/clawgame
```
Comment (missatessatessy) [inline code/modules/research/designs/comp_board_designs.dm]: even your own typepath that you added, NOT a wepon

File: code/game/objects/items/circuitboards/machine_circuitboards.dm [hunk 10]
```diff
@@ -1192,3 +1192,12 @@
 		/obj/item/stock_parts/manipulator = /obj/item/stock_parts/manipulator/femto,
 		/obj/item/stock_parts/micro_laser = /obj/item/stock_parts/micro_laser/quadultra,
 		/obj/item/stock_parts/scanning_module = /obj/item/stock_parts/scanning_module/triphasic)
+
+/obj/item/circuitboard/clawgame
```
Comment (missatessatessy) [inline code/game/objects/items/circuitboards/machine_circuitboards.dm]: /obj/item/circuitboard/machine/clawgame
dont forget to change this where it's used

File: code/modules/arcade/arcade_prize.dm [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+	icon_state = pick("prizeball_1","prizeball_2","prizeball_3")
+
+/obj/item/toy/prizeball/attack_self(mob/user as mob)
+	if(opening)
+		return
+	opening = 1
+	playsound(src.loc, 'sound/items/bubblewrap.ogg', 30, 1, extrarange = -4, falloff = 10)
+	icon_state = "prizeconfetti"
+	src.color = pick(random_color_list)
+	var/prize_inside = pick(/obj/random/carp_plushie, /obj/random/plushie, /obj/random/figure)	//will add ticket bundles later
```
Comment (missatessatessy) [inline code/modules/arcade/arcade_prize.dm]: these don't exist, so you'll probably have to make a list yourself unless you can find the yogs equivalent

File: code/modules/arcade/arcade_prize.dm [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+	..()
+	icon_state = pick("prizeball_1","prizeball_2","prizeball_3")
+
+/obj/item/toy/prizeball/attack_self(mob/user as mob)
+	if(opening)
+		return
+	opening = 1
+	playsound(src.loc, 'sound/items/bubblewrap.ogg', 30, 1, extrarange = -4, falloff = 10)
+	icon_state = "prizeconfetti"
+	src.color = pick(random_color_list)
```
Comment (missatessatessy) [inline code/modules/arcade/arcade_prize.dm]: also no color list that i can find

File: code/modules/arcade/arcade_prize.dm [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+	var/prize_inside = pick(/obj/random/carp_plushie, /obj/random/plushie, /obj/random/figure)	//will add ticket bundles later
+	spawn(10)
+		user.unEquip(src)
```
Comment (missatessatessy) [inline code/modules/arcade/arcade_prize.dm]: probably dont need this line, var doesnt exist anyway

File: code/modules/arcade/arcade_base.dm [hunk 14]
```diff
@@ -0,0 +1,169 @@
...
+/obj/machinery/arcade/update_icon()
+	return
+
+/obj/machinery/arcade/proc/pay_with_cash(var/obj/item/weapon/spacecash/cashmoney, var/mob/user)
+	if(cashmoney.get_total() < token_price)
+		user << "\icon[cashmoney] <span class='warning'>That is not enough money.</span>"
+		return 0
+	visible_message("<span class='info'>[usr] inserts a credit chip into [src].</span>")
+	var/left = cashmoney.get_total() - token_price
+	user.unEquip(cashmoney)
```
Comment (missatessatessy) [inline code/modules/arcade/arcade_base.dm]: unequip thingy again

File: code/modules/arcade/arcade_base.dm [hunk 15]
```diff
@@ -0,0 +1,169 @@
...
+	qdel(cashmoney)
+	if(left)
+		dispense_cash(left, src.loc, user)
```
Comment (missatessatessy) [inline code/modules/arcade/arcade_base.dm:97]: proc doesnt exist and I cba finding an equivalent right now

File: code/modules/arcade/claw_game.dm [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+	var/machine_image = "_1"
+	var/bonus_prize_chance = 5		//chance to dispense a SECOND prize if you win, increased by matter bin rating
+
+	//This is to make sure the images are available
+	var/list/img_resources = list('icons/obj/arcade_images/backgroundsprite.png',
+								'icons/obj/arcade_images/clawpieces.png',
+								'icons/obj/arcade_images/crane_bot.png',
+								'icons/obj/arcade_images/crane_top.png',
+								'icons/obj/arcade_images/prize_inside.png',
+								'icons/obj/arcade_images/prizeorbs.png')
```
Comment (missatessatessy) [inline code/modules/arcade/claw_game.dm]: you've used the wrong typepath for all of these that you actually added them as..

**Review by missatessatessy — COMMENTED** [review]
Here's some points for now, there's probably more that's wrong

**Comment by jcatjmeow** [comment]
This is a classic bood copypaste pr. Classic.

**Comment by jcatjmeow** [comment]
https://forums.yogstation.net/index.php?threads/a-guide-to-testing-your-prs-before-you-pr-them.22028/

**Pull request closed without merging**